### PR TITLE
Infostream

### PR DESCRIPTION
**Pull request by mcdee**
This implements the infostream streaming validator status APIs as per https://github.com/prysmaticlabs/ethereumapis/pull/96

### CODE REVIEW

File: beacon-chain/rpc/beacon/validators_stream.go [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+}
+
+// StreamValidatorsInfo returns a stream of information for given validators.
+// Validators are supplied dynamically by the client, and can be added, removed and reset at any time.
+func (bs *Server) StreamValidatorsInfo(stream ethpb.BeaconChain_StreamValidatorsInfoServer) error {
+	pubKeys := make([][]byte, 0)
+	pubKeysMutex := sync.RWMutex{}
+	stateChannel := make(chan *feed.Event, 1)
+	stateSub := bs.StateNotifier.StateFeed().Subscribe(stateChannel)
+	eth1Deposits := cache.New(time.Duration(params.BeaconConfig().SecondsPerSlot*params.BeaconConfig().SlotsPerEpoch)*time.Second, 12*time.Minute)
```
Comment (shayzluf) [inline beacon-chain/rpc/beacon/validators_stream.go]: maybe it will be better to name those vars and make prune time a multiple of epoch time

Comment (mcdee) [reply beacon-chain/rpc/beacon/validators_stream.go]: Thanks; named and tidied up.

File: beacon-chain/rpc/beacon/validators_stream.go [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+		case <-bs.Ctx.Done():
+			return status.Error(codes.Canceled, "Service context canceled")
+		case <-stream.Context().Done():
+			return status.Error(codes.Canceled, "Stream context canceled")
+		}
+	}
+}
+
+func (bs *Server) generateValidatorInfo(ctx context.Context, pubKeys [][]byte, eth1Deposits *cache.Cache, depositMutex *sync.RWMutex) ([]*ethpb.ValidatorInfo, error) {
+	headState, err := bs.HeadFetcher.HeadState(ctx)
```
Comment (nisdas) [inline beacon-chain/rpc/beacon/validators_stream.go]: this method is rather big, can we break it up ?

File: beacon-chain/rpc/beacon/validators_stream.go [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+							stream.Context().Done()
+							break
+						}
+					}
+				}
+			}
+		}
+	}()
+	// Send responses at the end of every epoch.
+	for {
```
[the remark this answers is not in the document]
Comment (rauljordan) [reply beacon-chain/rpc/beacon/validators_stream.go]: Agree, each of the cases could be a helper function if possible

File: beacon-chain/rpc/beacon/validators_stream.go [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+
+		res = append(res, info)
+
+		// Keep track of pending validators to fill in activation epoch later.
+		if info.Status == ethpb.ValidatorStatus_PENDING {
+			pendingValidatorsMap[bytesutil.ToBytes48(info.PublicKey)] = len(res) - 1
+		}
+	}
+
+	// Calculate activation time for pending validators (if there are any).
```
Comment (rauljordan) [inline beacon-chain/rpc/beacon/validators_stream.go]: This can be its own func

File: beacon-chain/rpc/beacon/validators_stream.go [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+		}
+
+		// Index
+		var ok bool
+		info.Index, ok, err = bs.BeaconDB.ValidatorIndex(ctx, pubKey)
+		if err != nil {
+			return nil, status.Error(codes.Internal, "Failed to obtain validator index")
+		}
+		if !ok {
+			// We don't know of this validator; perhaps it's a pending deposit?
```
Comment (rauljordan) [inline beacon-chain/rpc/beacon/validators_stream.go]: Make this a separate func

File: beacon-chain/core/helpers/validators.go [hunk 6]
```diff
@@ -270,6 +270,13 @@ func IsEligibleForActivationQueue(validator *ethpb.Validator) bool {
 		validator.EffectiveBalance == params.BeaconConfig().MaxEffectiveBalance
 }
 
+// IsEligibleForActivationQueueUsingTrie checks if the read-only validator is eligible to
+// be placed into the activation queue.
+func IsEligibleForActivationQueueUsingTrie(validator *stateTrie.ReadOnlyValidator) bool {
+	return validator.ActivationEligibilityEpoch() == params.BeaconConfig().FarFutureEpoch &&
```
[the remark this answers is not in the document]
Comment (mcdee) [reply beacon-chain/core/helpers/validators.go]: Broken the logic out in to its own function.  At current the function isn't exported; could be changed if there is a need but not sure having three separate paths to the same function (with validator, with read only validator, direct) is necessary.

File: beacon-chain/core/helpers/validators.go [hunk 7]
```diff
@@ -288,3 +295,13 @@ func IsEligibleForActivation(state *stateTrie.BeaconState, validator *ethpb.Vali
 	return validator.ActivationEligibilityEpoch <= finalizedEpoch &&
 		validator.ActivationEpoch == params.BeaconConfig().FarFutureEpoch
 }
+
+// IsEligibleForActivationUsingTrie checks if the validator is eligible for activation.
+func IsEligibleForActivationUsingTrie(state *stateTrie.BeaconState, validator *stateTrie.ReadOnlyValidator) bool {
```
[the remark this answers is not in the document]
Comment (mcdee) [reply beacon-chain/core/helpers/validators.go]: Ditto.

File: beacon-chain/rpc/beacon/validators_stream.go [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+			}
+			switch msg.Action {
+			case ethpb.SetAction_ADD_VALIDATOR_KEYS:
+				pubKeysMutex.Lock()
+				// Create existence map to ensure we don't duplicate keys.
+				pubKeysMap := make(map[[48]byte]bool)
+				for _, pubKey := range pubKeys {
+					pubKeysMap[bytesutil.ToBytes48(pubKey)] = true
+				}
+				addedPubKeys := make([][]byte, 0)
```
[the remark this answers is not in the document]
Comment (mcdee) [reply beacon-chain/rpc/beacon/validators_stream.go]: Added capacities to maps and slices where appropriate.

File: beacon-chain/rpc/beacon/validators_stream.go [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+				if validators, err := bs.generateValidatorInfo(bs.Ctx, addedPubKeys, eth1Deposits, depositMutex); err == nil {
+					for _, validator := range validators {
+						if err := stream.Send(validator); err != nil {
+							stream.Context().Done()
+							break
+						}
+					}
+				}
+			case ethpb.SetAction_REMOVE_VALIDATOR_KEYS:
+				msgPubKeysMap := make(map[[48]byte]bool)
```
[the remark this answers is not in the document]
Comment (mcdee) [reply beacon-chain/rpc/beacon/validators_stream.go]: Added capacities to maps and slices where appropriate.

File: beacon-chain/rpc/beacon/validators_stream.go [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+				}
+				pubKeysMutex.Unlock()
+			case ethpb.SetAction_SET_VALIDATOR_KEYS:
+				pubKeysMutex.Lock()
+				pubKeys = make([][]byte, 0, len(msg.PublicKeys))
+				for _, pubKey := range msg.PublicKeys {
+					pubKeys = append(pubKeys, pubKey)
+				}
+				pubKeysMutex.Unlock()
+				if validators, err := bs.generateValidatorInfo(bs.Ctx, msg.PublicKeys, eth1Deposits, depositMutex); err == nil {
```
Comment (0xKiwi) [inline beacon-chain/rpc/beacon/validators_stream.go]: Why is the `SET_VALIDATOR_KEYS` query sending the info streams?

File: beacon-chain/rpc/beacon/validators_stream.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mcdee) [reply beacon-chain/rpc/beacon/validators_stream.go]: Broken down in to separate functions.

File: beacon-chain/rpc/beacon/validators_stream.go (hunk 4 above)
Comment (mcdee) [reply beacon-chain/rpc/beacon/validators_stream.go]: Broken down in to separate functions.

File: beacon-chain/rpc/beacon/validators_stream.go (hunk 5 above)
Comment (mcdee) [reply beacon-chain/rpc/beacon/validators_stream.go]: Broken down in to separate functions.

File: beacon-chain/rpc/beacon/validators_stream.go (hunk 10 above)
Comment (mcdee) [reply beacon-chain/rpc/beacon/validators_stream.go]: Setting (or adding) the keys sends current data for the keys, then future updates as epochs tick past.

**Pull request merged**